File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l=0;
        int h=0;
        int max=0;
        Set<Character> st=new HashSet<>();
        while(h<s.length()){
            
            while(st.contains(s.charAt(h))){
                st.remove(s.charAt(l));
                l++;
            }
            st.add(s.charAt(h));
            max=Math.max(max,h-l+1);
            h++;
        }
        return max;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left=0
        ri=0
        maxlen=0
        st=set()
        while ri<len(s):
            c=s[ri]
            while c in st:
                st.remove(s[left])
                left+=1
            st.add(c)
            maxlen=max(maxlen,ri-left+1)
            ri+=1
        return maxlen